compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -1442,21 +1447,33 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
         val gadtCtx = ctx.fresh.setFreshGADTBounds.addMode(Mode.GadtConstraintInference)
         if (reducePattern(caseBindingMap, scrutineeSym.termRef, cdef.pat)(using gadtCtx)) {
           val (caseBindings, from, to) = substBindings(caseBindingMap.toList, mutable.ListBuffer(), Nil, Nil)
+          val typedGuard = typer.typed(cdef.guard.subst(from, to), defn.BooleanType)
```

Comment:
We should avoid type checking this if `cdef.guard.isEmpty`

Reply:
You could use somthing like this instead
```scala
val (guardOK, canReduceGuard) = 
  if cdef.guard.isEmpty then (true, true)
  else typer.typed(cdef.guard.subst(from, to), defn.BooleanType) match {
    case ConstantValue(v: Boolean) => (v, true)
    case _ => (false, false)
  }
```

Reply:
ok, done